Yank/storage/storageinterface.py

```diff
@@ -0,0 +1,514 @@
+            self._directory = self._storage_driver.get_directory(self.path, create=False)
+
+    def _write_append_directory(self):
+        """Special function of a predecessor to try and fetch the directory from file"""
+        self._storage_interface._instance_read()
+        self._check_directory()
+        if self._directory is None or self._directory is True:
+            self._directory = self._storage_driver.get_directory(self.path, create=True)
+        self._dump_metadata_buffer()
+
+    def __getattr__(self, name):
+        if self._variable:
+            raise AttributeError("Cannot convert this object to a directory as its already bound to a variable!")
+        if not self._directory:
+            # Assign directory features
+            self._directory = True
+        setattr(self, name, StorageInterfaceDirVar(name, self._storage_driver, self._storage_interface,
+                                                   predecessor=self))
+        return getattr(self, name)
+
+# =============================================================================
+# STORAGE INTERFACE
+# =============================================================================
+
+
+class StorageInterface(object):
+    """
+    This class holds what folders and variables are known to the file on the disk, and dynamically
+    creates them on the fly. Any attempt to reference a property which is not explicitly listed below implies that you
+    wish to define either a storage directory or variable, and creates a new StorageInterfaceVarDir (SIVD) with that
+    name. The SIVD is what handles the read/write operations on this disk by interfacing with the uninstanced
+    StorageIODriver you provide to storage_system keyword in StorageInterface class's __init__.
+
+    See StorageInterfaceVarDir for how the dynamic interfacing works.
+
+    Examples
+    --------
+
+    Create a basic storage system and write new my_data to disk
+    >>> my_data = [4,2,1,6]
+    >>> my_storage = StorageInterface('my_store.nc')
+    >>> my_storage.my_variable.write(my_data)
+
+    Create a folder called "vardir" with two variables in it, "var1" and "var2" holding DATA1 and DATA2 respectively
+    Then, fetch data from the same folders and variables as above
+    >>> DATA1 = "some string"
+    >>> DATA2 = (4.5, 7.0, -23.1)
+    >>> my_storage = StorageInterface('my_store.nc')
+    >>> my_storage.vardir.var1.write(DATA1)
+    >>> my_storage.vardir.var2.write(DATA2)
+    >>> var1 = my_storage.vardir.var1.fetch()
+    >>> var2 = my_storage.vardir.var2.fetch()
+
+    Run some_function() to generate data over an iteration, store each value in a dynamically sized var called "looper"
+    >>> mydata = [-1, 24, 5]
+    >>> my_storage = StorageInterface('my_store.nc')
+    >>> for i in range(10):
+    >>>     mydata = some_function()
+    >>>     my_storage.looper.append(mydata)
+    """
+    def __init__(self, file_name, storage_system=NetCDFIODriver):
```

Comment:
We just talked about this, but I'm saving this for the records. We should pass an object instead of a class name to make the user be able to pass options in the driver's constructor (including his own codecs).